### PR TITLE
Fix unaligned double word accesses in chunkset_c on ARMv7

### PR DESCRIPTION
**Pull request by NiLuJe**
c.f., https://github.com/zlib-ng/zlib-ng/issues/708#issuecomment-667622465

### CODE REVIEW

**Comment by NiLuJe** [comment]
~~Does not yet implement the recommendation from https://github.com/zlib-ng/zlib-ng/issues/708#issuecomment-667706892~~

File: chunkset.c [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #include "zbuild.h"
 #include "zutil.h"
 
-#ifdef UNALIGNED_OK
+#ifdef UNALIGNED64_OK
 typedef uint64_t chunk_t;
```
Comment (nmoinvaz) [inline chunkset.c:10]: We can do:
```
#if defined(UNALIGNED64_OK)
typedef uint64_t chunk_t;
#elif defined(UNALIGNED_OK)
typedef uint32_t chunk_t;
#else
typedef uint8_t chunk_t;
#endif
```

But the functions below would also need to support `chunk_t` being `uint32_t`.

**Comment by NiLuJe** [comment]
Re: https://github.com/zlib-ng/zlib-ng/issues/708#issuecomment-667706892

Would something like that work?

```diff
diff --git a/deflate.h b/deflate.h
index 153ec25..6e3d431 100644
--- a/deflate.h
+++ b/deflate.h
@@ -97,7 +97,7 @@ typedef struct tree_desc_s {
 typedef uint16_t Pos;
 
 /* A Pos is an index in the character window. We use short instead of int to
- * save space in the various tables. 
+ * save space in the various tables.
  */
 
 typedef struct internal_state {
@@ -355,14 +355,10 @@ static inline void put_uint64(deflate_state *s, uint64_t lld) {
     *(uint64_t *)(&s->pending_buf[s->pending]) = lld;
     s->pending += 8;
 #else
-    put_byte(s, (lld & 0xff));
-    put_byte(s, ((lld >> 8) & 0xff));
-    put_byte(s, ((lld >> 16) & 0xff));
-    put_byte(s, ((lld >> 24) & 0xff));
-    put_byte(s, ((lld >> 32) & 0xff));
-    put_byte(s, ((lld >> 40) & 0xff));
-    put_byte(s, ((lld >> 48) & 0xff));
-    put_byte(s, ((lld >> 56) & 0xff));
+    *(uint32_t *)(&s->pending_buf[s->pending]) = lld & 0xffffffff;
+    s->pending += 4;
+    *(uint32_t *)(&s->pending_buf[s->pending]) = (lld >> 32) & 0xffffffff;
+    s->pending += 4;
 #endif
 }
```

File: chunkset.c (hunk 1 above)
Comment (NiLuJe) [reply chunkset.c:10]: True, but I'd have to actually grok what's happening in there ^^.

I'll take a closer look at it after dinner, but no promises ;).

**Comment by nmoinvaz** [comment]
It looks right. But you would want something like this:
```
#if defined(UNALIGNED64_OK)
     *(uint64_t *)(&s->pending_buf[s->pending]) = lld;
     s->pending += 8;
#elif defined(UNALIGNED_OK)
     *(uint32_t *)(&s->pending_buf[s->pending]) = lld & 0xffffffff;
     s->pending += 4;
     *(uint32_t *)(&s->pending_buf[s->pending]) = (lld >> 32) & 0xffffffff;
     s->pending += 4;
#else
     put_byte(s, (lld & 0xff));
     put_byte(s, ((lld >> 8) & 0xff));
     put_byte(s, ((lld >> 16) & 0xff));
     put_byte(s, ((lld >> 24) & 0xff));
     put_byte(s, ((lld >> 32) & 0xff));
     put_byte(s, ((lld >> 40) & 0xff));
     put_byte(s, ((lld >> 48) & 0xff));
     put_byte(s, ((lld >> 56) & 0xff));
#endif
```

**Comment by NiLuJe** [comment]
@nmoinvaz: Oh, yeah, of course. :facepalm: 

EDIT: In 188c55e

File: chunkset.c (hunk 1 above)
Comment (NiLuJe) [reply chunkset.c:10]: Does that look okay? Not *quite* confident about `chunkmemset_4`, but the rest is trivial enough (maybe :D).

```diff
diff --git a/chunkset.c b/chunkset.c
index 121f3a4..423c69e 100644
--- a/chunkset.c
+++ b/chunkset.c
@@ -5,8 +5,10 @@
 #include "zbuild.h"
 #include "zutil.h"
 
-#ifdef UNALIGNED64_OK
+#if defined(UNALIGNED64_OK)
 typedef uint64_t chunk_t;
+#elif defined(UNALIGNED_OK)
+typedef uint32_t chunk_t;
 #else
 typedef uint8_t chunk_t[8];
 #endif
@@ -16,18 +18,24 @@ typedef uint8_t chunk_t[8];
 #define HAVE_CHUNKMEMSET_8
 
 static inline void chunkmemset_1(uint8_t *from, chunk_t *chunk) {
-#ifdef UNALIGNED64_OK
+#if defined(UNALIGNED64_OK)
     *chunk = 0x0101010101010101 * (uint8_t)*from;
+#elif defined(UNALIGNED_OK)
+    *chunk = 0x01010101 * (uint8_t)*from;
 #else
     memset(chunk, *from, sizeof(chunk_t));
 #endif
 }
 
 static inline void chunkmemset_4(uint8_t *from, chunk_t *chunk) {
-#ifdef UNALIGNED64_OK
+#if defined(UNALIGNED64_OK)
     uint32_t half_chunk;
     half_chunk = *(uint32_t *)from;
     *chunk = 0x0000000100000001 * (uint64_t)half_chunk;
+#elif defined(UNALIGNED_OK)
+    uint16_t half_chunk;
+    half_chunk = *(uint16_t *)from;
+    *chunk = 0x00010001 * (uint32_t)half_chunk;
 #else
     uint8_t *chunkptr = (uint8_t *)chunk;
     memcpy(chunkptr, from, 4);
@@ -36,8 +44,10 @@ static inline void chunkmemset_4(uint8_t *from, chunk_t *chunk) {
 }
 
 static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
-#ifdef UNALIGNED64_OK
+#if defined(UNALIGNED64_OK)
     *chunk = *(uint64_t *)from;
+#elif defined(UNALIGNED_OK)
+    *chunk = *(uint32_t *)from;
 #else
     memcpy(chunk, from, sizeof(chunk_t));
 #endif
@@ -48,8 +58,10 @@ static inline void loadchunk(uint8_t const *s, chunk_t *chunk) {
 }
 
 static inline void storechunk(uint8_t *out, chunk_t *chunk) {
-#ifdef UNALIGNED64_OK
+#if defined(UNALIGNED64_OK)
     *(uint64_t *)out = *chunk;
+#elif defined(UNALIGNED_OK)
+    *(uint32_t *)out = *chunk;
 #else
     /* Cast to chunk_t pointer to avoid compiler error on MSVC ARM */
     memcpy((chunk_t *)out, chunk, sizeof(chunk_t ));
```

A very quick bsdtar test on a random tarball appears successful with this, FWIW ;).

Comment (nmoinvaz) [reply chunkset.c:10]: In `chunkmemset_4`:
```
+#elif defined(UNALIGNED_OK)
+    uint16_t half_chunk;
+    half_chunk = *(uint16_t *)from;
+    *chunk = 0x00010001 * (uint32_t)half_chunk;
```
Should just be:
```
+#elif defined(UNALIGNED_OK)
+    *chunk = *(uint32_t *)from;
```
And for `chunkmemset_8`, I don't think it should exist if defined `UNALIGNED64_OK` is not defined and neither should be `HAVE_CHUNKMEMSET_8`. So wrap those in `#ifdef UNALIGNED64_OK`..

Comment (NiLuJe) [reply chunkset.c:10]: Should be taken care of w/ e09ef09, thanks for the pointers (ha!)!

**Comment by NiLuJe** [comment]
@nmoinvaz: Oops ^^. I think I got everything in 6516925. It builds, and bsdtar still doesn't implode ;).

**Review by nmoinvaz — APPROVED** [review]
LGTM

File: chunkset.c [hunk 2]
```diff
@@ -36,8 +44,12 @@ static inline void chunkmemset_4(uint8_t *from, chunk_t *chunk) {
 }
 
 static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
-#ifdef UNALIGNED_OK
+#if defined(UNALIGNED64_OK)
     *chunk = *(uint64_t *)from;
+#elif defined(UNALIGNED_OK)
+    uint64_t full_chunk = *(uint64_t *)from;
```
Comment (mtl1979) [inline chunkset.c]: This is still possibly 64-bit unaligned read.

Comment (NiLuJe) [reply chunkset.c]: The fact that it's *declared* as an uint64_t on the stack *should* mean it's properly aligned, though, right?

Comment (NiLuJe) [reply chunkset.c]: Oh, wait, the *read*, not the store. :D. :facepalm:

Comment (mtl1979) [reply chunkset.c]: The problem is ```from``` which is not guaranteed to be on stack and even if it is on stack, it might not be aligned on 64-bit boundary. Stack can be aligned on either 32-bit or 64-bit boundary depending on system.

On release builds, optimizer might not generate any code for this line, but debug builds must preserve all local variables that are used.

Comment (NiLuJe) [reply chunkset.c]: Although, `from` itself is an ` uint8_t *`, so it's *probably* never going to be an issue?

(My brain hurts >_<).

In any case, GCC appears to be chunking that in two 32bits loads anyway:

```
3980   │ static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
3981   │ #if defined(UNALIGNED64_OK)
3982   │     *chunk = *(uint64_t *)from;
3983   │ #elif defined(UNALIGNED_OK)
3984   │     uint64_t full_chunk = *(uint64_t *)from;
3985   │     3bb2:   682c        ldr r4, [r5, #0]
3986   │     3bb4:   686b        ldr r3, [r5, #4]
```

Comment (NiLuJe) [reply chunkset.c]: I guess I can make that explicit by doing it this way instead (i.e., relying on good old pointer arithmetic rules)?

```C
    uint32_t* p = (uint32_t *)from;
    *chunk[0] = *p;
    *chunk[1] = *(p+1);
```

```
3979   │ static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
3980   │ #if defined(UNALIGNED64_OK)
3981   │     *chunk = *(uint64_t *)from;
3982   │ #elif defined(UNALIGNED_OK)
3983   │     uint32_t* p = (uint32_t *)from;
3984   │     *chunk[0] = *p;
3985   │     3bb2:   682c        ldr r4, [r5, #0]
3986   │     *chunk[1] = *(p+1);
3987   │     3bb4:   686b        ldr r3, [r5, #4]
```

Comment (mtl1979) [reply chunkset.c]: Being ```uint8_t *``` only implies that the pointer is assumed to be unaligned. When you cast it to wider type, you essentially create temporary variable that must be assigned to another variable with same alignment restrictions.... Basically you can only assign the value of memory it references to, to vector of ```uint8_t``` or another memory location pointed by another variable of type ```uint8_t *```. Assigning values explicitly to registers was removed from recent gcc versions, even though compiler itself might temporarily use registers to store the variable contents.

Basically this code relies on compiler generating the correct code instead of writing it as two 32-bit copies explicitly.

Comment (mtl1979) [reply chunkset.c]: Or even...

```
uint32_t* p = (uint32_t *)from;
chunk[0] = p[0];
chunk[1] = p[1];
```

Comment (NiLuJe) [reply chunkset.c]: This has a no impact on register usage once inlined in `chunkcopy_c`:

```
4066   │     uint64_t full_chunk = *(uint64_t *)from;
4067   │     3c0e:   680c        ldr r4, [r1, #0]
4068   │     3c10:   684b        ldr r3, [r1, #4]
4069   │     ((uint32_t *)out)[0] = *chunk[0];
4070   │     3c12:   6004        str r4, [r0, #0]
4071   │     ((uint32_t *)out)[1] = *chunk[1];
4072   │     3c14:   6043        str r3, [r0, #4]
```

```
4067   │     *chunk[0] = *p;
4068   │     3c0e:   680c        ldr r4, [r1, #0]
4069   │     *chunk[1] = *(p+1);
4070   │     3c10:   684b        ldr r3, [r1, #4]
4071   │     ((uint32_t *)out)[0] = *chunk[0];
4072   │     3c12:   6004        str r4, [r0, #0]
4073   │     ((uint32_t *)out)[1] = *chunk[1];
4074   │     3c14:   6043        str r3, [r0, #4]
```

Comment (NiLuJe) [reply chunkset.c]: And a minor one (for the better?) in `CHUNKMEMSET`:

```
6140   │     uint64_t full_chunk = *(uint64_t *)from;
6141   │     4af8:   682a        ldr r2, [r5, #0]
6142   │     4afa:   686b        ldr r3, [r5, #4]
6143   │     *chunk[0] = (uint32_t)(full_chunk & 0xffffffff);
6144   │     4afc:   60ba        str r2, [r7, #8]
6145   │     *chunk[1] = (uint32_t)(full_chunk >> 32);
6146   │     4afe:   613b        str r3, [r7, #16]
```

```
6143   │     *chunk[0] = *p;
6144   │     4af8:   682b        ldr r3, [r5, #0]
6145   │     4afa:   60bb        str r3, [r7, #8]
6146   │     *chunk[1] = *(p+1);
6147   │     4afc:   686b        ldr r3, [r5, #4]
6148   │     4afe:   613b        str r3, [r7, #16]
```

Comment (NiLuJe) [reply chunkset.c]: @mtl1979: Gah, I missed your comments, but it looks like we had mostly the same idea at around the same time ;).

Comment (NiLuJe) [reply chunkset.c]: Now that I'm caught up, is there a preference towards the array index notation vs. the pointer arithmetic one?

Comment (mtl1979) [reply chunkset.c]: When one of the variables is type[], then it makes more sense to use array notation on both sides even if the other variable is type* as it is more consistent and easier to read. Pointer arithmetics should only be used when all variables are type*, or when using operator++ etc. to advance the pointer itself.

File: chunkset.c [hunk 3]
```diff
@@ -5,8 +5,10 @@
...
 #include "zutil.h"
 
-#ifdef UNALIGNED_OK
+#if defined(UNALIGNED64_OK)
 typedef uint64_t chunk_t;
+#elif defined(UNALIGNED_OK)
+typedef uint32_t chunk_t[2];
 #else
 typedef uint8_t chunk_t[8];
 #endif
```
Comment (mtl1979) [inline chunkset.c:15]: This is completely wrong... size of chunk_t is always correct, but the type in function prototype/parameter list is not... Mixing arrays with non-arrays in typedef makes the code confusing and prone to errors.

We don't want uint32_t*[2] as the function parameter type. we want uint32_t* when UNALIGNED64_OK not defined.
We don't want uint8_t*[8] as the function parameter type. we want uint8_t* when neither UNALIGNED_OK or UNALIGNED64_OK is defined..

File: chunkset.c [hunk 4]
```diff
@@ -5,12 +5,13 @@
 #include "zbuild.h"
 #include "zutil.h"
 
+// We need sizeof(chunk_t) to be 8, no matter what.
 #if defined(UNALIGNED64_OK)
 typedef uint64_t chunk_t;
 #elif defined(UNALIGNED_OK)
-typedef uint32_t chunk_t[2];
+typedef struct { uint32_t u32[2]; } chunk_t;
```
Comment (mtl1979) [inline chunkset.c]: If we really want to support pre-C11 compilers, then the start of the line should be modified as "typedef struct chunk_t {"... so the struct itself has same name as the variable type the typedef declares.

Same below ;)

**Review by nmoinvaz — APPROVED** [review]
Looks good.

File: chunkset.c (hunk 4 above)
Comment (Dead2) [reply chunkset.c]: We have stated that we target C99 compilers, so we should make sure we keep it compatible.
I don't see anything really interesting (for us) in C11 anyways.

Comment (mtl1979) [reply chunkset.c]: That's why I commented how this particular line should be modified -- here -- instead of continuing the discussion on the original location. Sometimes long conversations start to wander around and it becomes hard to follow which line(s) are being discussed.

Comment (NiLuJe) [reply chunkset.c]: Done! (Sorry, was away on vacation ;)).

**Comment by Dead2** [comment]
Looks good to me.
